### ISSUE TITLE
[BUG] slow compilation by using val in a local block for the same variable name

### ISSUE THREAD

**Issue by mschaaf** [rzwitserloot/lombok#2652]
**Describe the bug**
We have a gradle, spring, java build and use lombok in our test and main code.
We run into a compilation performance issue with using `val`.
We introduced the following test method and the compilation takes now 10 seconds longer
Slow to compile:
```java
    @SneakyThrows
    @Test
    void testImport() {
        {
            val user = ValueStore.user();
            user.setEmail("");
            user.setActive(true);
            userRepository.save(user);
        }

        try (val pkg = OPCPackage.open(TestClass.class.getResourceAsStream("/file.xlsx"))) {
            try (val wb = new XSSFWorkbook(pkg)) {
                val sheet = wb.getSheetAt(1);

                val connection = DriverManager.getConnection(url, username, password);
                val jdbcConnection = new JdbcConnection(connection);
                val mysqlDatabase = new MySQLDatabase();
                mysqlDatabase.setConnection(jdbcConnection);

                val consumer = new Consumer(mysqlDatabase);
                consumer.accept(sheet.getRow(0));
                consumer.accept(sheet.getRow(1));
                consumer.accept(sheet.getRow(2));

                mysqlDatabase.commit();
                mysqlDatabase.close();

                Assertions.assertThat(consumer.getUsersCreated().intValue()).isEqualTo(1);
                Assertions.assertThat(consumer.getUsersUpdated().intValue()).isEqualTo(1);
            }
        }

        {
            val user = userRepository.findByEmail("a");
            Assertions.assertThat(user.getFirstname()).isEqualTo("A");
            Assertions.assertThat(user.getLastname()).isEqualTo("C");
            Assertions.assertThat(user.getEmail()).isEqualTo("a");
            Assertions.assertThat(user.getEmail2()).isNull();
            Assertions.assertThat(user.getEmail3()).isNull();
            Assertions.assertThat(user.getActive()).isFalse();
            Assertions.assertThat(userRepository.getEmployeeId("a")).isEqualTo(0);
        }

        {
            val user = userRepository.findByEmail("b");
            Assertions.assertThat(user.getFirstname()).isNotEqualTo("A");
            Assertions.assertThat(user.getLastname()).isNotEqualTo("G");
            Assertions.assertThat(user.getEmail()).isEqualTo("b");
            Assertions.assertThat(user.getEmail2()).isNull();
            Assertions.assertThat(user.getEmail3()).isNull();
            Assertions.assertThat(user.getActive()).isTrue();
            Assertions.assertThat(userRepository.getEmployeeId("b")).isEqualTo(0);
        }

    }
```

When the method was changed to replace `val user` with `final User` the compilation is again at the speed as without the method.
Fast to compile:
```java
    @SneakyThrows
    @Test
    void testImport() {
        {
            final User user = ValueStore.user();
            user.setEmail("");
            user.setActive(true);
            userRepository.save(user);
        }

        try (val pkg = OPCPackage.open(TestClass.class.getResourceAsStream("/file.xlsx"))) {
            try (val wb = new XSSFWorkbook(pkg)) {
                val sheet = wb.getSheetAt(1);

                val connection = DriverManager.getConnection(url, username, password);
                val jdbcConnection = new JdbcConnection(connection);
                val mysqlDatabase = new MySQLDatabase();
                mysqlDatabase.setConnection(jdbcConnection);

                val consumer = new Consumer(mysqlDatabase);
                consumer.accept(sheet.getRow(0));
                consumer.accept(sheet.getRow(1));
                consumer.accept(sheet.getRow(2));

                mysqlDatabase.commit();
                mysqlDatabase.close();

                Assertions.assertThat(consumer.getUsersCreated().intValue()).isEqualTo(1);
                Assertions.assertThat(consumer.getUsersUpdated().intValue()).isEqualTo(1);
            }
        }

        {
            final User user = userRepository.findByEmail("a");
            Assertions.assertThat(user.getFirstname()).isEqualTo("A");
            Assertions.assertThat(user.getLastname()).isEqualTo("C");
            Assertions.assertThat(user.getEmail()).isEqualTo("a");
            Assertions.assertThat(user.getEmail2()).isNull();
            Assertions.assertThat(user.getEmail3()).isNull();
            Assertions.assertThat(user.getActive()).isFalse();
            Assertions.assertThat(userRepository.getEmployeeId("a")).isEqualTo(0);
        }

        {
            final User user = userRepository.findByEmail("b");
            Assertions.assertThat(user.getFirstname()).isNotEqualTo("A");
            Assertions.assertThat(user.getLastname()).isNotEqualTo("G");
            Assertions.assertThat(user.getEmail()).isEqualTo("b");
            Assertions.assertThat(user.getEmail2()).isNull();
            Assertions.assertThat(user.getEmail3()).isNull();
            Assertions.assertThat(user.getActive()).isTrue();
            Assertions.assertThat(userRepository.getEmployeeId("b")).isEqualTo(0);
        }

    }
```

Replacing the other `val` constructs in this method seems to have no effect. Only the use of `val user` seems to increase the compile time.
In other methods with much more of such `val` constructs for the same variable name, we see compilation increases by more than two minutes.

The whole compilation takes from 10 minutes on a fast machine to 20 minutes on a sower machine.
By removing some of these `val` constructs we could reduce the compile time by about 5 minutes and it seems we can reduce it further.

We would like to use `val` all over the place but with this performance breakdown, we have to replace `val`.

**To Reproduce**
For now, we don't have a simple project to share that reproduces this behaviour.

**Expected behavior**
The use of `val` should not slow down the compilation so much, because it is one of the most used annotations in our project.

**Version info (please complete the following information):**
 - Lombok version: 1.18.16
 - Platform gradle 4.7.1, java: 11.0.9 (Eclipse OpenJ9 openj9-0.23.0), os: linux

**Additional context**

We would like to help to identify the issue.
Is there a way to profile the execution of lombok during compilation?
Are there other ways to identify what slows down the usage of `val` in this context.

**Comment by Rawi01** [comment]
Seems to be similar to #2131

**Comment by rzwitserloot** [comment]
Well, sounds like a more or less self  contained test case so somebody can profile this properly. Thanks, @mschaaf

**Comment by mschaaf** [comment]
a simplified flamegraph of the slowed down compilation:
[flames.svg.gz](https://github.com/rzwitserloot/lombok/files/5747844/flames.svg.gz)

**Comment by mschaaf** [comment]
a raw flamegraph of the slowed down compilation:
[flames.svg.gz](https://github.com/rzwitserloot/lombok/files/5747850/flames.svg.gz)

**Comment by mschaaf** [comment]
There are millions of InapplicableMethodExceptions and FunctionDescriptorLookupError exceptions thrown during resolution:

```
com.sun.tools.javac.code.Types$DescriptorCache.findDescriptorInternal(Symbol$TypeSymbol, Scope$CompoundScope) Types.java:735
com.sun.tools.javac.comp.Attr.attribStat(JCTree, Env) Attr.java:724
lombok.javac.JavacResolution.attrib(JCTree, Env) JavacResolution.java:235
lombok.javac.JavacResolution.memberEnterAndAttribute(JCTree, Env, Context) JavacResolution.java:203
lombok.javac.JavacResolution.resolveMethodMember(JavacNode) JavacResolution.java:151
lombok.javac.handlers.HandleVal.visitLocal(JavacNode, JCTree$JCVariableDecl) HandleVal.java:133 3367800
lombok.javac.JavacNode.traverse(JavacASTVisitor) JavacNode.java:120
...
```

**Comment by rzwitserloot** [comment]
millions? Oof. `val` is a tricky beast. This is a bug and we should fix it, but it feels like it's not an easy fix.

Is your team on jdk11+? I wonder if we should detect java version and define `val` behaviour as just turning itself into `final var` if the host platform AND the `-source` option (which I think we can figure out, though that'll require some extra code to the framework layer) allow it (so, JDK10+).

One problem is that `final var` and `lombok.val` are almost, but not quite, equivalent:

* `var` allows you to have unwritable types, such as LUBs (`var x = list.get(0);` where `list` boils down to `List<Foobar & Serializable>` - is legal, and makes `x` be of type `Foobar & Serializable`, even though you can't actually write that in a java file. All methods of both Foobar and Serializable are invokable. An even easier test; `var x = new Object() { void foo() {} }; x.foo();` is legal, but there's nothing you can write for `var`, other than `var`, to make that work. `val x = new Object() { void foo() {} }; x.foo();` does not work, because according to the rules of `val`, the best 'writable' type is Object, which has no foo method.
* However, this is legal: `lombok.val x = 10, y = "Hello";` whereas `var` disallows that comma entirely. Actually, it looks like `val` is broken in this regard.

The primary issue is that you write code that works if compiled on JDK11 with `-source 10` or higher, and breaks if you don't, but not with an error that states: "Hey, you are using a feature that is not supported at this source level".

That's a cost, but probably worth paying; even if we make `val` a lot faster, it remains a bit of a hairy, resolution-based feature, and replacing it with `var` is much simpler. Unfortunately, this may not be as simple as I'm making it out to be, because `javac` "unrolls" a var quite early in the process last I checked, it's possible lombok runs 'too late'. But, before I investigate if we can do that... good idea, or no? If it can be done at all it's probably a much, much faster path to avoiding the performance drain than endeavouring to fix the problem properly.

**Comment by mschaaf** [comment]
We are now at java 15 and at the time of reporting, we were on java 11.

**Comment by rzwitserloot** [comment]
> We are now at java 15 and at the time of reporting, we were on java 11.

I think I can conclude from this you'd prefer the much quicker fix of turning `val` into `final var`? (potentially quicker - not 100% sure we can do this, but if we can, it'll probably be easy to implement)

**Comment by mschaaf** [comment]
Yes, I prefer a quick fix because it would allow us to continue without making an exception in our code for `val` until it is fixed.
I can test Lombok versions with a potential fix if necessary.

**Comment by Rawi01** [comment]
I managed to reproduce the issue and simplified the test case:
```java
import org.assertj.core.api.Assertions;
import lombok.val;

class ValMissingImport {
	public void test() {
		val obj = 1;
		Assertions.assertThat(obj);
		Assertions.assertThat(obj);
		Assertions.assertThat(obj);
		Assertions.assertThat(obj);
		Assertions.assertThat(obj);
		Assertions.assertThat(obj);
	}
}
```
Every additional call of `Assertions.assertThat(obj);` increases compilation time by ~2 seconds. Resolving the type of `obj` is no problem but I guess that if fails trying to resolve `___Lombok_VAL_Attrib__` while trying to figure out which overloaded `assertThat` is the right one.

Simply replacing `lombok.val` with `final var` seems to be the easiest solution. To do that we just have to set `vartype` to `null`. Unfortunately this delomboks to `final var` and breaks all `var`/`val` related tests.

**Comment by mschaaf** [comment]
@Rawi01 could you please show me what code I need to change to achieve this or create a pull request with the changes?
I would like to build a version to speed up our compilation time.

**Comment by Rawi01** [comment]
@mschaaf Pushed the [test branch](https://github.com/Rawi01/lombok/tree/val-to-final-var) and also uploaded a [pre-built version](https://github.com/rzwitserloot/lombok/files/5909801/lombok.zip) for your tests.

**Comment by mschaaf** [comment]
@Rawi01 Thanks for sharing the code and the pre-built version. I included it in our project and the test folder compilation is now down to 5s from several minutes. Thanks a lot.

**Comment by otbe** [comment]
I think we're also affected by this but we're still on java 8 (corretto). Only tests seem to be affected (20s vs 2m30 execution time). Would be nice to have a fix for this :)

**Comment by mschaaf** [comment]
Is there a timeline for a release with this fix?

**Comment by Rawi01** [comment]
@otbe This is the only report for Java 8, can you double check that you really use the Java 8 **compiler**? 
@mschaaf There are some minor incompatibilities between `val` and `final var` and my test branch was simply a proof of concept. I just tried if the replacement is possible in eclipse too and it seems to work so if @rzwitserloot approves that concept I can adjust the test cases and create a PR.

**Comment by coffeemoney** [comment]
Hi @Rawi01,
we are having same issue (integration test running very slow with val/var) in java 8.

**Comment by Rawi01** [comment]
@coffeemoney Runtime or compilation time? Lombok is gone at runtime so there should be no slow down compared to normal non-lombok code. Does that mean that it takes 131 minutes to compile your test? :fearful: I'm also curious how you measured that time difference :smile:

**Comment by otbe** [comment]
Yes @Rawi01 we're using java8 compiler. Replacing lomboks `val` by explicit types in tests reduces compilation time by 2minutes (2m30s vs 20s). Non test code does not seem to be affected.

**Comment by Rawi01** [comment]
@otbe Can you share a test case that helps me to reproduce the issue? Otherwise I can only guess what the problem might be with Java 8.

**Comment by Rawi01** [comment]
@otbe I pushed a [new branch](https://github.com/Rawi01/lombok/tree/var-val-slow) and also uploaded a [pre-built version](https://github.com/rzwitserloot/lombok/files/6316083/lombok.zip), can you try if that works properly and if it reduces the compilation time?

**Comment by otbe** [comment]
Hi @Rawi01 
Sorry for my late response. I tested your version (just copied over the lombok.jar) but unfortunately the compiler complains about the usage of `val` 

```
 Cannot use 'val' here because initializer expression does not have a representable type: Type cannot be resolved
```

**Comment by Rawi01** [comment]
Meeh, thats sad... Seems like the hacky solution I added can not handle all cases... Can you share one of the classes that fails to compile?

**Comment by mschaaf** [comment]
Is there a chance to get the fix for Java >11 in the main code? We would like to avoid maintaining our own special copy of lombok.

**Comment by chrisgleissner** [comment]
Just to provide some context of how dramatic the impact of this bug is, by replacing ca. 200 `val` references with their `var` counterparts in ca. 70 Java files, I was able to reduce the build-time of a Java 11 project from 2m 30s to 12s. Latest Lombok version.

We are using Gradle 7 and compile with the latest OpenJDK 11. The bug affected both compilation of main and test code, i.e. speedup for both was comparable. Until this issue is fixed, I can't recommend using `val` for any sizeable project.

I see there have been no comments since May 28th. Before, there was some discussion of simply replacing `val` with `final var`. Did it turn out that this is not technically possible and what is the way forward on this bug?

Thanks

**Comment by Rawi01** [comment]
If I remeber correctly I almost finished the missing eclipse part and the test cases. I will try to create a PR this week for the `val` to `final var` replacement.
